### PR TITLE
Better compatibility with custom AuthProviders

### PR DESCRIPTION
**Pull request by leoafarias**
There are some cases with custom AuthProvider that might choose to return something on the promise.
You can see an example here:
https://github.com/benwinding/react-admin-firebase/blob/master/src/providers/AuthProvider.ts#L198

By returning void, it makes these providers incompatible and display type errors.

### CODE REVIEW

File: packages/ra-core/src/types.ts [hunk 1]
```diff
@@ -68,8 +68,8 @@ export interface UserIdentity {
 export type AuthProvider = {
     login: (params: any) => Promise<any>;
     logout: (params: any) => Promise<void | string>;
-    checkAuth: (params: any) => Promise<void>;
-    checkError: (error: any) => Promise<void>;
+    checkAuth: (params: any) => Promise<any>;
+    checkError: (error: any) => Promise<any>;
```
Comment (djhi) [inline packages/ra-core/src/types.ts:72]: Then a generic type with a default to `void` might be better don't you think?

**Review by djhi — CHANGES_REQUESTED** [review]
Thanks :)

File: packages/ra-core/src/types.ts (hunk 1 above)
Comment (djhi) [reply packages/ra-core/src/types.ts:72]: Somehing like:
```ts
checkError: <T = void>(error: any) => Promise<T>;
```

Comment (leoafarias) [reply packages/ra-core/src/types.ts:72]: @djhi I tried the following initially. However, I still get a type error when adding the authProvider
```typescript
checkError: <T = any>(error: any) => Promise<T>;
```

I get the following linting error the same applies for `void`
```
Type 'Promise<User>' is not assignable to type 'Promise<T>'.
        Type 'User' is not assignable to type 'T'.
          'T' could be instantiated with an arbitrary type which could be unrelated to 'User'
```

I could get around this by doing the following.
```tsx
 <Admin
        dataProvider={dataProvider}
        authProvider={authProvider as AuthProvider}
      >
```

It's your call, but just some AuthProviders might need to have to be overriden.

Let me know what you think.

Comment (leoafarias) [reply packages/ra-core/src/types.ts:72]: At least the override would work. With `Promise<void>` the override errors out saying the types are incompatible.

File: packages/ra-core/src/types.ts [hunk 2]
```diff
@@ -68,8 +68,8 @@ export interface UserIdentity {
 export type AuthProvider = {
     login: (params: any) => Promise<any>;
     logout: (params: any) => Promise<void | false | string>;
-    checkAuth: (params: any) => Promise<void>;
-    checkError: (error: any) => Promise<void>;
+    checkAuth: <T = any>(error: any) => Promise<T>;
```
Comment (fzaninotto) [inline packages/ra-core/src/types.ts:71]: `checkAuth` param shouldn't be named `error`, it's misleading

**Comment by fzaninotto** [comment]
TypeScript doesn't seem to like your changes...

**Comment by leoafarias** [comment]
> TypeScript doesn't seem to like your changes...

Yes, that is the reason why my initial commit did not have a generic.

**Comment by fzaninotto** [comment]
We won't move further with. this PR: It's the authPRovider's responsibility to respect the contract.

**Pull request closed without merging**